Dockerfile

```diff
@@ -1,32 +1,46 @@
-FROM registry.access.redhat.com/ubi9/python-311@sha256:39bd10c0d80a127280095ff352f8a6881e61c39c2edbbdf5206e6508adc3b8e7 AS prod
-COPY --from=ghcr.io/astral-sh/uv:0.4.26@sha256:7775c60dca9cc5827c36757c32c75985244d8f31447565fa8147e2b2e11ad280 /uv /bin/uv
+#
+# Builder image
+# Keep version in sync with prod image below
+#
+FROM registry.access.redhat.com/ubi9/python-311@sha256:129391e5b291e29a3b24a3fdb281a56d7cb8297bdf8ba92bb5c8581b53084324 AS builder
+COPY --from=ghcr.io/astral-sh/uv:0.4.29@sha256:ebb10c5178c7a357d80527f3371e7038561c26234e8a0bb323ea1f2ce8a694b7 /uv /bin/uv
 
 # er-outputs-secrets version. keep in sync with pyproject.toml
-LABEL konflux.additional-tags="0.2.0"
+LABEL konflux.additional-tags="0.2.1"
 
 ENV \
     # use venv from ubi image
     UV_PROJECT_ENVIRONMENT="/opt/app-root" \
     # compile bytecode for faster startup
     UV_COMPILE_BYTECODE="true" \
     # disable uv cache. it doesn't make sense in a container
-    UV_NO_CACHE=true \
-    # uv will run without updating the uv.lock file.
-    UV_FROZEN=true
+    UV_NO_CACHE=true
 
 COPY LICENSE /licenses/MIT
 
 # Install dependencies
 COPY pyproject.toml uv.lock ./
-RUN uv sync --no-install-project --no-dev
+RUN uv sync --frozen --no-install-project --no-group dev
 
 COPY main.py ./
-
+RUN uv sync --frozen --no-group dev
+#
+# Production image
+#
+FROM registry.access.redhat.com/ubi9/python-311@sha256:129391e5b291e29a3b24a3fdb281a56d7cb8297bdf8ba92bb5c8581b53084324 AS prod
+COPY --from=builder /opt/app-root /opt/app-root
 ENTRYPOINT [ "python3", "main.py" ]
 
+#
+# Test image
+#
 FROM prod AS test
+COPY --from=ghcr.io/astral-sh/uv:0.4.29@sha256:ebb10c5178c7a357d80527f3371e7038561c26234e8a0bb323ea1f2ce8a694b7 /uv /bin/uv
```

Comment:
if `test` is from `builder`, then some duplications can be removed like this


```dockerfile
FROM builder AS test
RUN uv sync --frozen
COPY Makefile ./
COPY tests ./tests
RUN make test
```

Better to put `prod` as the last stage in `Dockerfile`, so when no target specified, it will by default build the last stage, and `test` stage is auto skipped.

Reply:
Changed in 4a9222a